ddsketch/src/lib.rs

```diff
@@ -0,0 +1,461 @@
+            .collect();
+        bins.sort_by(|a, b| a.0.total_cmp(&b.0));
+        bins
+    }
+
+    /// Add a point with value `point` to the sketch
+    pub fn add(&mut self, point: f64) -> Option<()> {
+        self.add_with_count(point, 1.0)
+    }
+
+    /// Add `count` point with value `point` to the sketch
+    pub fn add_with_count(&mut self, point: f64, count: f64) -> Option<()> {
+        if count.is_nan() || count.is_infinite() {
+            return None;
+        }
+        if point < 0.0 || point.is_nan() || point.is_infinite() {
+            return None;
+        } else if point < self.mapping.min_indexable_value {
+            self.zero_count += count;
+        } else {
+            let index = self.mapping.index(point);
+            *self.store.bin_mut(index) += count;
+        }
+        Some(())
+    }
+
+    /// Return a protobuf of the sketch
+    pub fn into_pb(self) -> pb::DdSketch {
+        let contiguous_bins: Vec<f64> = self.store.bins.into();
+        pb::DdSketch {
+            mapping: Some(pb::IndexMapping {
+                gamma: self.mapping.gamma,
+                index_offset: self.mapping.index_offset,
+                interpolation: pb::index_mapping::Interpolation::None.into(),
+            }),
+            positive_values: Some(pb::Store {
+                bin_counts: HashMap::new(),
+                contiguous_bin_counts: contiguous_bins,
+                contiguous_bin_index_offset: self.store.offset,
+            }),
+            zero_count: self.zero_count,
+            negative_values: Some(pb::Store {
+                bin_counts: HashMap::new(),
+                contiguous_bin_counts: Vec::new(),
+                contiguous_bin_index_offset: 0,
+            }),
+        }
+    }
+
+    /// Return a serialized protobuf of the sketch
+    pub fn encode_to_vec(self) -> Vec<u8> {
+        self.into_pb().encode_to_vec()
+    }
+}
+
+/// A store mapping the bin indexes to their respective weights
+///
+/// Stores the weights as contiguousBinCounts, only the bins within `offset` and the highest
+/// non empty bin are stored.
+///
+/// Stores the weights of a contiguous range of bins containing all non-empty bins. The range start
+/// at index `offset` and end at index `offset + bins.len()`
+///
+/// The range of stored bins is updated when accessing the index of a bin out of the range
+/// with [`Self::bin_idx_to_bin_idx()`]. If the `max_size` is reached the lower bins are
+/// collapsed together to free space.
+#[derive(Debug, Clone)]
+struct LowCollapsingDenseStore {
+    bins: VecDeque<f64>,
+    offset: i32,
+    max_size: i32,
+}
+
+impl LowCollapsingDenseStore {
+    fn new(max_size: i32) -> Option<Self> {
+        if max_size < 0 {
+            return None;
+        }
+        Some(Self {
+            bins: VecDeque::new(),
+            offset: 0,
+            max_size,
+        })
+    }
+
+    /// Return an iterator over the bins
```

Comment:
nitpick: I prefer to use regular comments over rustdoc comments for private functions. You wouldn't want this exposed if you generate docs.

Reply:
Private functions aren't printed in rustdocs I think

Reply:
Ah, I just tried it and you're correct. It still comes through in IDEs, which is nice.

Reply:
You can also use `--document-private-items` to generate docs with it.